Meetings iframe visibility

entantoencuanto opened the pull request:
#### :tophat: What? Why?

This PR:

* Adds a new enum attribute to meetings to control level access to iframe
* Adds management in meetings admin panel to select iframe access level when iframe is enabled for meeting
* Changes online meeting cell visibility considering:
  * Iframe access level of meeting
  * Privacy and transparency of meeting

#### :pushpin: Related Issues

- Fixes #7719 
- Fixes #7720 

#### Testing

1. Sign in as admin
2. In the dashboard, go to edit a meeting form in a process
3. Select "Online" as meeting type
4. See that you have the options "Iframe embed type" and "Iframe access level". They should work as described at #7719 


#### :clipboard: Checklist
:rotating_light: Please review the [guidelines for contributing](https://github.com/decidim/decidim/blob/develop/CONTRIBUTING.adoc) to this repository.

- [ ] :question: **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] :heavy_check_mark: **DO** check open PR's to avoid duplicates.
- [ ] :heavy_check_mark: **DO** keep pull requests small so they can be easily reviewed.
- [ ] :heavy_check_mark: **DO** build locally before pushing.
- [ ] :heavy_check_mark: **DO** make sure tests pass.
- [ ] :heavy_check_mark: **DO** make sure any new changes are documented in `docs/`.
- [ ] :heavy_check_mark: **DO** add and modify seeds if necessary.
- [ ] :heavy_check_mark: **DO** add CHANGELOG upgrade notes if required.
- [ ] :heavy_check_mark: **DO** add to GraphQL API if there are new public fields.
- [ ] :heavy_check_mark: **DO** add link to MetaDecidim if it's a new feature.
- [ ] :x:**AVOID** breaking the continuous integration build.
- [ ] :x:**AVOID** making significant changes to the overall architecture.

### :camera: Screenshots

![image](https://user-images.githubusercontent.com/717367/140710580-d00af677-b480-4080-a33b-be0353c4858e.png)


#### Embed type: embed in meeting page

![image](https://user-images.githubusercontent.com/717367/140710658-b18fb54e-4a28-49a2-bdf7-50b11ba675ca.png)


#### Embed type: live event

![image](https://user-images.githubusercontent.com/717367/140710689-7d43c9dc-365e-4b27-b0c6-242fd4fbc2f2.png)


#### Embed type: external

![image](https://user-images.githubusercontent.com/717367/140710710-a26fde9e-a7ea-415b-9f80-21b2ffd096e3.png)


:hearts: Thank you!

andreslucena commented:
I think that we actually have some missing pieces in this puzzle, mainly we're missing the /live_event interface for an embedded type of meetings. One of the main use cases that we have in Decidim Barcelona is using Online Meetings with Jitsi and/or 8x8 or similar platforms (with WebRTC), and with this final PR, this isn't supported. This is closely related to the "Live Event" interface developed in #8065.

At the moment if I configure a Meeting with
- Type: Online
- Online Meeting URL: https://meet.jit.si/decidim-testing
Then when I click in "Join Meeting" I go to the external link directly, but I need some way of configuring it to the "Live Event" UI. 

So, to have this we'd need to have a couple of changes: 

1. Remove the "Show embedded iframe for this URL" checkbox
2. Create a new dropdown selector called "Iframe embed type" with options:
2.1 None: it doesn't do anything (no button, no live_event access). It's the default option. When this is selected "Iframe access level" is hidden.
2.2 Embed in meeting page : current behavior as the checkbox checked, so it'd be migrated from the meetings that have already checked "Show embedded iframe for this URL"
2.3 Open in live event page (with optional polls): the button would link to /live_event URL. ACLs for seeing the button and accessing /live_event URL would be honored by the "Iframe access level" setting in both cases. 
2.4 Open in a new tab: the button would link to the external link (current behavior of checkbox "Show embedded iframe for this URL" unchecked)
3. Regarding "Embed in meeting page" (old "Show embedded iframe for this URL") we'd add the "meet.jit.si" URL to the whitelist. It doesn't need any transformation. In the future, we should be adding other services that people use to this whitelist, although I don't know if they need any kind of transformation (ie 8x8.com or Whereby.com), so for the moment I'd only add Jitsi. 

As a summary:

<table>
<tr>
	<td>Options
	<td>Button
	<td>Link
	<td>Explanation
<tr>
	<td>None
	<td>No
	<td>No
	<td>Do nothing
<tr>
	<td>Embed in meeting page
	<td>No
	<td>No
	<td>Shows the iframe in the same page
<tr>
	<td>Open in live event page (with optional polls)
	<td>Yes
	<td>Yes, link to /live_event
	<td>Shows the iframe in the live event page
<tr>
	<td>Open in a new tab
	<td>Yes
	<td>Yes, link to Online Meeting URL
	<td>Opens the URL in a new tab
</table>

This is how it'd finally look like: 
 
![imatge](https://user-images.githubusercontent.com/717367/133429752-c9470adf-084d-4e4a-9c7c-7fb4a1c7377f.png)
![imatge](https://user-images.githubusercontent.com/717367/133429872-eb294f7a-2eab-4c88-8416-cef7c5208fcc.png)

entantoencuanto commented:
I have started working on the changes proposed

cc/ @decidim/product

entantoencuanto commented:
I've implemented the changes and also:
* Added jitsi to the allowlist of embeddable in iframe URLs
* Included a validation to reject meetings creation/updates when the URL doesn't fit in the embeddable allowlist and the iframe embed type is set as "Open in live event page (with optional polls)" (this validation also works if "Embed in meeting page" is set)

Ready to check in the test application @decidim/product

andreslucena commented:
A couple of bugs I've found: 

- [ ] Given that I'm a visitor, and there's a meeting configured with "Open in live event page (with optional polls)" and "Only signed-in participants" then I see the "Link available soon" in the card (I shouldn't see it, it should be consistent with the meeting page): 

![imatge](https://user-images.githubusercontent.com/717367/135219531-e70f4b6b-0ac1-4fc0-842c-0a884df56f5a.png)

- [ ] Given that I'm a visitor, and there's a meeting configured with "Open in a new tab" and "Only signed-in participants" then I see the "Link available soon" in the card (I shouldn't see it, it should be consistent with the meeting page): 

![imatge](https://user-images.githubusercontent.com/717367/135219601-4a533795-6836-4677-8fa2-65a04d07b4b6.png)

- [ ] Given that there's a meeting configured with "Embed in meeting page" and the start date is a couple of months from now (for instance, 20220125), then I see the embed (I shouldn't see it, it should honour timing from #7722)

![imatge](https://user-images.githubusercontent.com/717367/135222952-5ae48ac3-1efb-491e-9d98-b84025317a00.png)

- [ ] In the admin form we should change the help text to better reflect these changes:

> Only a few services allow embedding in meeting or live event (YouTube, Twitch and Jitsi)


![imatge](https://user-images.githubusercontent.com/717367/135223920-099dbde7-d164-4052-a820-65eb915b4345.png)

entantoencuanto commented:
Hi, @decidim/product , I've added the requested changes and It's ready to test in the staging application

andreslucena commented on decidim-meetings/config/locales/en.yml:
```diff
@@ -310,7 +310,9 @@ en:
-            show_embedded_iframe_help: Whether or not embed the iframe for this videoconference URL. Only a few services allow embedding (i.e. Youtube, Twitch...)
+            select_an_iframe_access_level: Please select an iframe access level
+            select_an_iframe_embed_type: Please select an iframe embed type
+            show_embedded_iframe_help: Only a few services allow embedding (i.e. Youtube, Twitch...)
```
```suggestion
            show_embedded_iframe_help: Only a few services allow embedding in meeting or live event (YouTube, Twitch and Jitsi)
```

andreslucena replied:
This wasn't accepted. Can you accept it @entantoencuanto? Thanks

andreslucena commented on decidim-meetings/config/locales/en.yml:
```diff
@@ -482,7 +493,8 @@ en:
           select_a_registration_type: Please select a registration type
-          show_embedded_iframe_help: Whether or not embed the iframe for this videoconference URL. Only a few services allow embedding (i.e. Youtube, Twitch...)
+          select_an_iframe_embed_type: Please select an iframe embed type
+          show_embedded_iframe_help: Only a few services allow embedding (i.e. Youtube, Twitch...)
```
```suggestion
          show_embedded_iframe_help: Only a few services allow embedding in meeting or live event (YouTube, Twitch and Jitsi)
```

andreslucena commented:
> Hi, @decidim/product , I've added the requested changes and It's ready to test in the staging application

Awesome, I've reviewed it, we're getting closer!!

I've updated the description and Acceptance Criteria of #7719 so it's possible to follow with the last changes. A couple more details: 

1. I've left some [review comments](https://github.com/decidim/decidim/pull/8307#pullrequestreview-771310071) changing the help text so it's clearer 

Regarding the updated Acceptance Ctiteria: 

2. There's one about not showing ACL if it doesn't make sense:

> - [ ] _Given that I'm an administrator configuring an online meeting when I select "Iframe embed type" of "None" then I should not see the "Iframe access level" dropdown_

3. The other ones are related to the /live_event URL and a possible attack. If a user knows this URL then she can access it directly without honoring the defined ACL. She can know it because this is open source, she has already accessed this URL for another meeting and can infer it, she has received it from another participant that has access, etc... We'd need to follow the same ACL as the "show button" logic (that works as expected and defined!!). In this case, the ACLs that aren't working are: 

>   - [ ]  _Given that there's a Meeting with videoconference when it's configured with "Iframe embed type" of "Open in live event page (with optional polls)" and with "Iframe access level" of "Only signed-in participants" when I'm not logged in then I should not see the button and I should not be able to visit /live_event URL_
>  - [ ]  _Given that there's a Meeting with videoconference when it's configured with "Iframe embed type" of "Open in live event page (with optional polls)" and with "Iframe access level" of "Registered participants to this meeting" when I'm not logged in then I should not see the button and I should not be able to visit /live_event URL_
>  - [ ]   _Given that there's a Meeting with videoconference when it's configured with "Iframe embed type" of "Open in live event page (with optional polls)" and with "Iframe access level" of "Registered participants to this meeting when I'm logged in and not registered to this meeting then I should not see the button and I should not be able to visit /live_event URL_

Could you check this @entantoencuanto 🙏🏽  ?

We're really close!! 🚀

andreslucena commented:
> The iframe access level is now available also when "Open URL in a new tab" is selected. I've also deployed again and I think that it's working as expected now 

Great, now it works as expected with this dropdown!! With that we can say that #7719 is fixed. I've also reviewed the acceptance criterias defined in #7720 and is fixed too. 🥳 🎉 

But I see a problem with timing and  that the "live event":

- [ ] Given that there's a meeting configured with "Open in live event page (with optional polls)" and the start date is a couple of months from now (for instance, 20220125), then I can visit the /live_event URL (I shouldn't see it, it should honour timing from #7722)

And also it's missing this one: 

- [ ] In the admin form we should change the help text to better reflect these changes:

> Only a few services allow embedding in meeting or live event (YouTube, Twitch and Jitsi)

Regarding this last one, I've made the change myself. Please [accept this](https://github.com/decidim/decidim/pull/8307#discussion_r722141309)

entantoencuanto commented:
> Regarding this last one, I've made the change myself. Please [accept this](https://github.com/decidim/decidim/pull/8307#discussion_r722141309)

Sorry, I thought it was already accepted, I must have done something wrong. I think now it's merged correctly

entantoencuanto commented:
Ready to check @andreslucena 

cc/ @decidim/product

andreslucena commented:
@entantoencuanto everything works like a charm, congratulations!! 🥳 🥳 

I've added the "Testing" and "Screenshots" sections in the PR body description, so it's easy to follow to others without having to read our ping pong. Feel free to improve it.

andreslucena commented:
@entantoencuanto could you resolve the conflicts 🙏🏽 ? 

Also, I found out while reviewing this that we didn't take into account what happens with meetings created by participants: 

![image](https://user-images.githubusercontent.com/717367/143561491-f5da720e-e206-4e03-ac97-df1af4869977.png)

I think it'll be better to just handle this as another issue, as this PR is for sure an improvement on what we have until now. The definition would be to have what we have in the backend without the "live poll" option (as that's not implemented for meetings created by participants)

entantoencuanto commented:
Conflicts solved by rebasing

andreslucena commented on decidim-meetings/spec/system/live_meeting_access_spec.rb:
```diff
@@ -20,16 +20,216 @@ def visit_meeting
+            login_as user, scope: :user
+          end
+
+          it "shows the meeting link embedded" do
```
I don't like the "meeting link embedded" as a concept, but then I realized it makes perfect sense with the "iframe_embed_type" of "open_in_new_tab" 🙈 🤷🏽

andreslucena commented:
Great job in this huge PR @entantoencuanto - the internal joke that I always make in @decidim/product when talking about these features is that this is going to be awesome for the next pandemia 🤣 

> Also, I found out while reviewing this that we didn't take into account what happens with meetings created by participants:
> 
> ![image](https://user-images.githubusercontent.com/717367/143561491-f5da720e-e206-4e03-ac97-df1af4869977.png)
> 
> I think it'll be better to just handle this as another issue, as this PR is for sure an improvement on what we have until now. The definition would be to have what we have in the backend without the "live poll" option (as that's not implemented for meetings created by participants)

I've opened this issue in #8570 - it'd be great to have that when you have time @entantoencuanto. This will be the killer feature for v0.26.0, so as soon as we have that solved we'll release with that
